### ISSUE TITLE
enum C type conversion is wrong?

### ISSUE THREAD

**Issue by luke-jr** [thepowersgang/mrustc#30]
`rustc-1.19.0-src/src/vendor/log/src/lib.rs` contains:

```Rust
#[repr(usize)]
#[derive(Copy, Eq, Debug)]
pub enum LogLevelFilter {
```

mrustc seems to define `usize` as `uintptr_t`, but mrustc converts this to:

```C
struct e__ZN10log$$0_3_714LogLevelFilter {
    unsigned int TAG;
};
```

`usize` is correctly converted to `uintptr_t` later on, in `::"log-0_3_7"::max_log_level`, resulting in a `memcpy` of 64 bits (`uintptr_t` size on 64-bit platforms) to a data structure that is only `unsigned int` (32-bit).

Disclaimer: I don't know Rust, so double-check my assumptions.

**Comment by thepowersgang** [comment]
Confirmed - Low priority until bootstrap completed or causes incorrect behavior..

**Comment by luke-jr** [comment]
I would expect this to cause memory corruption and therefore crashing. Wouldn't it need to be fixed for bootstrap to work?

(`max_log_level` is only one example.. GCC was throwing warnings about this in many locations.)

**Comment by luke-jr** [comment]
In fact, the built rustc dies exactly because of this:

```
$ output/rustc helloworld.rs 
*** buffer overflow detected ***
Killed
```

```
$ valgrind output/rustc helloworld.rs 
==8540== Memcheck, a memory error detector
==8540== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==8540== Using Valgrind-3.12.0 and LibVEX; rerun with -h for copyright info
==8540== Command: output/rustc helloworld.rs
==8540== 
==8540== warning: Can't handle line info entry with line number 2313583 greater than 1048575
==8540== (Nb: this message is only shown once)
==8540== warning: Can't handle inlined call info entry with line number 1784829 greater than 1048575
==8540== (Nb: this message is only shown once)
**8540** *** memcpy_chk: buffer overflow detected ***: program terminated
==8540==    at 0x6174E54: VALGRIND_PRINTF_BACKTRACE (valgrind.h:6818)
==8540==    by 0x6179A7A: __memcpy_chk (vg_replace_strmem.c:1574)
==8540==    by 0x10B2045: memcpy (string3.h:53)
==8540==    by 0x10B2045: _ZN10log$$0_3_713max_log_level (librustc_metadata.hir.o.c:317054)
==8540==    by 0x11BEB3C: _ZRI$aL_ZN14rustc_metadata7creader11CrateLoader$aRresolve_crate (librustc_metadata.hir.o.c:453366)
==8540==    by 0x11C1E36: _ZRK$aL_ZN14rustc_metadata7creader11CrateLoader_as__ZN5rustc6middle6cstore11CrateLoader$aRprocess_item (librustc_metadata.hir.o.c:923917)
==8540==    by 0x6EB72F: _ZRI$aL_ZN13rustc_resolve8Resolver$aRbuild_reduced_graph_for_item (librustc_resolve.hir.o.c:219425)
==8540==    by 0x6ED7FE: _ZRK$aL_ZN13rustc_resolve19build_reduced_graph24BuildReducedGraphVisitor_as__ZN6syntax5visit7Visitor$aRvisit_item (librustc_resolve.hir.o.c:397423)
==8540==    by 0x6EDCD2: _ZN6syntax5visit8walk_mod$aL_ZN13rustc_resolve19build_reduced_graph24BuildReducedGraphVisitor$aR (librustc_resolve.hir.o.c:151055)
==8540==    by 0x6EDCD2: _ZN6syntax5visit9walk_item$aL_ZN13rustc_resolve19build_reduced_graph24BuildReducedGraphVisitor$aR (librustc_resolve.hir.o.c:149644)
==8540==    by 0x6EDCD2: _ZRK$aL_ZN13rustc_resolve19build_reduced_graph24BuildReducedGraphVisitor_as__ZN6syntax5visit7Visitor$aRvisit_item (librustc_resolve.hir.o.c:397427)
==8540==    by 0x6EFDFA: _ZRI$aL_ZN6syntax3ext6expand9Expansion$aRvisit_with$aL_ZN13rustc_resolve19build_reduced_graph24BuildReducedGraphVisitor$aR (librustc_resolve.hir.o.c:325626)
==8540==    by 0x6EFDFA: _ZRK$aL_ZN13rustc_resolve8Resolver_as__ZN6syntax3ext4base8Resolver$aRvisit_expansion (librustc_resolve.hir.o.c:395391)
==8540==    by 0x94597B: _ZRI$aL_ZN6syntax3ext6expand13MacroExpander$aRcollect_invocations (libsyntax.hir.o.c:391370)
==8540==    by 0x945AC7: _ZRI$aL_ZN6syntax3ext6expand13MacroExpander$aRexpand (libsyntax.hir.o.c:391520)
==8540==    by 0x94886C: _ZRI$aL_ZN6syntax3ext6expand13MacroExpander$aRexpand_crate (libsyntax.hir.o.c:393089)
==8540== 
==8540== HEAP SUMMARY:
==8540==     in use at exit: 439,988 bytes in 2,329 blocks
==8540==   total heap usage: 3,695 allocs, 1,366 frees, 639,268 bytes allocated
==8540== 
==8540== LEAK SUMMARY:
==8540==    definitely lost: 10,575 bytes in 39 blocks
==8540==    indirectly lost: 8,192 bytes in 13 blocks
==8540==      possibly lost: 304 bytes in 1 blocks
==8540==    still reachable: 420,917 bytes in 2,276 blocks
==8540==         suppressed: 0 bytes in 0 blocks
==8540== Rerun with --leak-check=full to see details of leaked memory
==8540== 
==8540== For counts of detected and suppressed errors, rerun with: -v
==8540== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by thepowersgang** [comment]
Should be fixed as part of cb271f8ea98d1a5c65a5e636a0e73a85710027b4 (which cleaned up a lot of the enum handling).

**Comment by thepowersgang** [comment]
Correction, fixed in 30aa97ca3d479719b78cbaaa3a1327c66d98de0c (will close once fix is confirmed)

**Comment by thepowersgang** [comment]
Closing, fix checked locally. If not actually fixed, please reopen.
